docs/architecture/adr-061-core-module-api.md

```diff
@@ -0,0 +1,562 @@
+interacting with genesis data represented by single `proto.Message` types.
+
+```go
+type GenesisSource interface {
+	ReadMessage(proto.Message) error
+	OpenReader(field string) (io.ReadCloser, error)
+	ReadRawJSON() (json.RawMessage, error)
+}
+
+type GenesisTarget interface {
+    WriteMessage(proto.Message) error
+    OpenWriter(field string) (io.WriteCloser, error)
+    WriteRawJSON(json.RawMessage) error
+}
+```
+
+All genesis objects for a given module are expected to conform to the semantics of a JSON object. That object may
+represent a single genesis state `proto.Message` which is the case for most current modules, or it may represent a set
+of nested JSON structures which may optionally be streamed as arrays from individual files which is the way the
+[ORM](./adr-055-orm.md) is designed. This streaming genesis support may be beneficial for chains with extremely large
+amounts of state that can't fit into memory.
+
+Modules which use a single `proto.Message` can use the `GenesisSource.ReadMessage` and `GenesisTarget.ReadMessage`
+methods without needing to deal with raw JSON or codecs.
+
+Modules which use streaming genesis can use the
+`GenesisSource.OpenReader` and `GenesisTarget.OpenWriter` methods. In the case of modules using the ORM, all JSON
+handling is done automatically by the ORM and a single line will be sufficient for registering all of that functionality
+(ex. `ormDb.RegisterGenesis(handler)`).
+
+Low-level `ReadRawJSON` and `WriteRawJSON` are provided in case modules use some other sort of legacy genesis logic
+we haven't anticipated.
+
+#### Begin and End Blockers
+
+The `BeginBlock` and `EndBlock` methods will simply take a `context.Context`, because:
+* most modules don't need Tendermint information other than `BlockInfo` so we can eliminate dependencies on specific
+Tendermint versions
+* for the few modules that need Tendermint block headers and/or return validator updates, specific versions of the
+runtime module will provide specific functionality for interacting with the specific version(s) of Tendermint
+supported
+
+In order for `BeginBlock`, `EndBlock` and `InitGenesis` to send back validator updates and retrieve full Tendermint
+block headers, the runtime module for a specific version of Tendermint could provide services like this:
+```go
+type ValidatorUpdateService interface {
+	SetValidatorUpdates(context.Context, []abci.ValidatorUpdate)
+}
+
+type BeginBlockService interface {
+	GetBeginBlockRequest(context.Context) abci.RequestBeginBlock 
+}
+```
+
+We know these types will change at the Tendermint level and that also a very limited set of modules actually need this
+functionality, so they are intentionally kept out of core to keep core limited to the necessary, minimal set of stable
+APIs.
+
+#### Event Listeners
+
+Handlers allow event listeners for typed events to be registered that will be called deterministically whenever those
```

Comment:
what about legacy events?

Reply:
This design doesn't allow for listening to legacy events, just typed events. My reasoning is that those events weren't formalized in a way to make them relevant to state machine consumers, so we should assume they're intended for clients only

Reply:
I also agree limiting state-machine events to only typed events is desirable. It can also be seen as an extra motivation to migrate to typed events asap (pending perf fixes).

Reply:
will runtime still support legacy events?

Reply:
Yes, there is an `EmitLegacy` method so that these types of events are supported for perpetuity